apps/docs/pages/guides/auth/auth-postgres-hooks.mdx

```diff
@@ -0,0 +1,669 @@
+    />
+
+
+
+## Use Cases
+
+Based on user feedback, we decided on a few use cases to illustrate how a hook. These use cases are not exhaustive and you can opt to [write your own hook](#writing-your-own-hook) if you have a use case which doesn’t fit any of the examples below.
+
+## MFA Verification Attempt Hook
+
+You may wish to increase security beyond the requirements of the [Supabase MFA implementation](https://supabase.com/docs/guides/auth/auth-mfa). You may need to meet compliance requirements set out by your client or you might be working in a field where data is sensitive.
+
+For such scenarios, the MFA Verification Attempt hook can allow you to add additional checks to:
+
+- Ask users to wait a set time between verification attempts, to limit the number of attempts that can be done in a period of time.
+- Force sign out users if they have too many invalid verification attempts.
+- Count, rate limit, ban or do other actions.
+
+### Inputs
+
+Supabase Auth will supply the following fields to your hook:
+
+- `factor_id` Unique identifier for the MFA factor the user is verifying. Correlate this to the `auth.mfa_factors` table.
+- `user_id` Unique identifier for the user. Correlate this with `auth.users`
+- `valid` Whether the verification attempt was valid, in the case for TOTP this means that the six digit code was correct (true) or incorrect (false). The hook will generally receive true once, and false many times for the same user and factor.
+
+Example payload:
+
+```json
+{
+	"factor_id": "6eab6a69-7766-48bf-95d8-bd8f606894db",
+	"user_id": "3919cb6e-4215-4478-a960-6d3454326cec",
+	"valid": true,
+}
+```
+
+### Outputs
+
+In exchange your hook should return back a JSON object with these fields set:
+
+********************************On Success********************************
+
+Return these only if your hook processed the input without errors.
+
+- `decision` A string containing the decision on whether to allow authentication to move forward. Use `reject` to completely reject the verification attempt and log the user out of all active sessions. Use `continue` to use the default Supabase Auth behavior.
+- `message` The message to show the user if the decision was `reject`.
+
+Example output:
+
+```json
+{
+   "decision": "reject",
+   "message" : "You have exceeded maximum number of MFA attempts."
+}
+```
+
+****************On Error****************
+
+Return this if your hook processed the input with some error.
+
+- `error` An object that contains information about the error.
+    - `http_code` A number indicating the HTTP code to be returned. If not set, the code is HTTP 500 Internal Server Error.
+    - `message` A message to be returned in the HTTP response. Must be set.
+- In general all other fields attached to the returned JSON are ignored.
+
+Example output:
+
+```jsx
+{
+  "error": {
+    "http_code": 429,
+    "message": "You can only verify a factor once very 10 seconds."
+  }
+}
+```
+## Password Verification Attempts Hook
```

Comment:
Add a warning here that explains the following:

Because this hook runs on _unauthenticated_ requests, this function can be called many times by just about anyone. Be careful when using this hook to prevent users from entering your app, as you can accidentally create a vulnerability where attackers can use this feature from preventing legitimate users to enter.

General recommendations:

- Always check if the password is valid first, before taking any additional action. This is a sure signal that the user is legitimate.
- Instead of blocking further password verification attempts, you can use this function to trigger an email or other notification to the user or administrator in your system.

---

In that direction, can you adjust the example to reflect these recommendations?